org.eclipse.lsp4e/src/org/eclipse/lsp4e/Versioned.java

```diff
@@ -12,32 +12,35 @@
  *******************************************************************************/
 package org.eclipse.lsp4e;
 
+import org.eclipse.jface.text.IDocument;
+import org.eclipse.lsp4e.internal.DocumentUtil;
+
 /**
  * Bundles together a result from a language server with the document version at the time it was run.
  * Supports optimistic locking.
  *
  * @param <T>
  */
 public class Versioned<T> {
-	private final long version;
-
-	private final T data;
+	protected final IDocument document;
+	protected final long sourceDocumentVersion;
+	protected final T data;
 
-	public Versioned(final long version, final T data) {
-		this.version = version;
+	/**
+	 * Constructs a new Versioned for the given document, and specify source version
+	 */
+	public Versioned(final IDocument document, long sourceDocumentVersion, final T data) {
+		this.document = document;
+		this.sourceDocumentVersion = sourceDocumentVersion;
 		this.data = data;
 	}
 
-	public T get() {
```

Comment:
I would rather keep the getters instead and make them protected than making the fields protected and removing the methods, but since the fields are final, it does not matter much in the end.

Reply:
I think "Effective Java" recommends usage of public (or protected here is enough) final fields for non-API objects. Adding getters and setters can make sense when things become more complex, but usually directly accessing final fields is pretty comfortable to uses and allows the JVM to run much more optimization.

Reply:
Thanks, I will check that. I checked the record classes in Java 17 in case we could use them, but we cannot because they are final.